source/grid/manifold_lib.cc

```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
+  boost::container::small_vector<double, 100>              distances(surrounding_points.size(),0.0);
+  double max_distance = 0.;
+  for (unsigned int i = 0; i < surrounding_points.size(); ++i)
+    {
+      directions[i] = surrounding_points[i] - center;
+      distances[i] = directions[i].norm();
+
+      if (distances[i] != 0.)
+        directions[i] /= distances[i];
+      else
+        Assert(false,
+               ExcMessage("One of the vertices coincides with the center. "
+                          "This is not allowed!"));
+
+      // Check if an estimate is good enough,
+      // this is often the case for sufficiently refined meshes.
+      for (unsigned int k = 0; k < i; ++k)
+        {
+          const double squared_distance = (directions[i] - directions[k]).norm_square();
+          max_distance = std::max(max_distance, squared_distance);
+        }
+    }
+
+  // Step 1: Check for some special cases, create simple linear guesses otherwise.
+  const double tolerance = 1e-10;
+  std::vector<bool> accurate_point_was_found(new_points.size(),false);
+  const ArrayView<const Tensor<1,spacedim>> array_directions = make_array_view(directions.begin(),directions.end());
+  const ArrayView<const double> array_distances = make_array_view(distances.begin(),distances.end());
+  for (unsigned int row=0; row<weights.size(0); ++row)
+    {
+      const std::pair<double, Tensor<1,spacedim>> candidate = guess_new_point(array_directions,
+                                                              array_distances,
+                                                              make_array_view(weights,row));
+
+      // If the candidate is the center, mark it as found to avoid entering
+      // the Newton iteration in step 2, which would crash.
+      if (candidate.first == 0.0)
+        {
+          new_points[row] = center;
+          accurate_point_was_found[row] = true;
+          continue;
+        }
+
+      // If not in 3D, just use the implementation from PolarManifold
+      // after we verified that the candidate is not the center.
+      if (spacedim<3)
+        {
+          new_points[row] = polar_manifold.get_new_point(surrounding_points, make_array_view(weights,row));
+          accurate_point_was_found[row] = true;
+          continue;
+        }
+
+      new_points[row] = center + candidate.first * candidate.second;
+    }
+
+  // If all the points are close to each other, we expect the estimate to
+  // be good enough. This tolerance was chosen such that the first iteration
+  // for a at least three time refined HyperShell mesh with radii .5 and 1.
+  // doesn't already succeed.
+  if (max_distance < 1e-2 || spacedim<3)
+    return;
+
+  // Step 2:
+  // Do more expensive Newton-style iterations to improve the estimate.
+
+  // Search for duplicate directions and merge them to minimize the cost of
+  // the get_new_point function call below.
+  Table<2,double> merged_weights(weights);
```

Comment:
It would be great if we could avoid this memory allocation. We could use a small vector (of size say 400) here as well and manually index into the array by `size1*row+i` and the like...